src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -461,6 +461,16 @@ not_ (ECompositeKey _) = throw (CompositeKeyErr NotError)
 (*.)  :: PersistField a => SqlExpr (Value a) -> SqlExpr (Value a) -> SqlExpr (Value a)
 (*.)  = unsafeSqlBinOp " * "
 
+-- | @BETWEEN@.
+--
+-- /Since: 3.0.0/
+between :: PersistField a => SqlExpr (Value a) -> (SqlExpr (Value a), SqlExpr (Value a)) -> SqlExpr (Value Bool)
+a `between` (ERaw p1 f, ERaw p2 g) = unsafeSqlBinOp " BETWEEN " a $ ERaw Never $ \x ->
+  let (v1, fv) = f x
+      (v2, gv) = g x
+  in  (parensM p1 v1 <> " AND " <> parensM p2 v2, fv ++ gv)
+_ `between` _                      = throw $ CompositeKeyErr BetweenError
```

Comment:
I think that we can properly handle this instead of throwing an error. `between` in Postgresql handles composite values just fine:

```sql
[local] matt@matt=# select ((5, 3) between (3, 2) and (6, 4));
 ?column? 
----------
 t
(1 row)
```

as does sqlite:

```sql
sqlite> select ((5, 3) between (3, 2) and (6, 4));
1
```

However, it doesn't work for MySQL -_-

```sql
mysql> select ((5, 3) between (3, 2) and (6, 4));
ERROR 1241 (21000): Operand should contain 1 column(s)
```

Sigh. This is the trouble with being generic over various SQL backends.

Especially annoying is the fact that MySQL works with the "dumb" implementation:

```sql
mysql> select ((5, 3) > (3, 2) and (5, 3) < (6, 4));
+---------------------------------------+
| ((5, 3) > (3, 2) and (5, 3) < (6, 4)) |
+---------------------------------------+
|                                     1 |
+---------------------------------------+
1 row in set (0.00 sec)
```

So, we have a few options:

1. Define `between` that uses the SQL keyword `between` with composite keys and reexport fromt he Postgresql and Sqlite modules. Define `between` that uses the expanded form for MySQL and export from the Mysql module.
2. Define `between` that uses `between` for scalar values and the expanded form for composite values. We'd need to ensure that it has the same semantics on the edges/boundaries.

Reply:
I didn't know about this, thanks @parsonsmatt will look into this also :smile:

Reply:
It could also be simply defined as:

```haskell
between a (x, y) = a >. x &&. a <. y
```

which should work for all backends. Not sure if that exactly matches the semantics of `BETWEEN` though.

Reply:
Maybe if it makes a mixture of option 2 and your last comment?
How about to make it work as it is now for scalar values and use `a >. x && a <.y` for composite keys?

Reply:
That works for me!

We need to ensure that we use the right operators - I'm not exactly sure what happens if you do `(1, 2) BETWEEN (1, 2) AND (2,3)` or `(2, 3) BETWEEN (1, 2) AND (2, 3)`.

Reply:
I will test this and get back to you!